File: Documentation/5.2/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/Querying/HowToWorkWithSuggestions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Threading.Tasks;
using Raven.Client.Documents;
using Raven.Client.Documents.Indexes;
using Raven.Client.Documents.Queries.Suggestions;
using Raven.Documentation.Samples.Orders;

namespace Raven.Documentation.Samples.ClientApi.Session.Querying
{
    public class HowToWorkWithSuggestions
    {
        public async Task Sample()
        {
            using (var store = new DocumentStore())
            {
                using (var session = store.OpenSession())
                {
                    #region suggest_1
                    // This dynamic query on the 'Products' collection has NO resulting documents
                    var product = session
                        .Query<Product>()
                        .Where(x => x.Name == "chaig")
                        .ToList();
                    #endregion
                }
                
                using (var session = store.OpenSession())
                {
                    #region suggest_2
                    // Query for suggested terms for single term:
                    // ==========================================
                    
                    Dictionary<string, SuggestionResult> suggestions = session
                         // Make a dynamic query on collection 'Products'
                        .Query<Product>()
                         // Call 'SuggestUsing'
                        .SuggestUsing(builder => builder
                             // Request to get terms from field 'Name' that are similar to 'chaig' 
                            .ByField(x => x.Name, "chaig"))
                        .Execute();
                    #endregion
                }
                
                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region suggest_3
                    // Query for suggested terms for single term:
                    // ==========================================
                    
                    Dictionary<string, SuggestionResult> suggestions = await asyncSession
                         // Make a dynamic query on collection 'Products'
                        .Query<Product>()
                         // Call 'SuggestUsing'
                        .SuggestUsing(builder => builder
                             // Request to get terms from field 'Name' that are similar to 'chaig'  
                            .ByField(x => x.Name, "chaig"))
                        .ExecuteAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region suggest_4
                    // Define the suggestion request for single term
                    var suggestionRequest = new SuggestionWithTerm("Name")
                    {
                        // Looking for terms from field 'Name' that are similar to term 'chaig'  
                        Term = "chaig"
                    };
                    
                    // Query for suggestions 
                    Dictionary<string, SuggestionResult> suggestions = session
                        .Query<Product>()
                         // Call 'SuggestUsing' - pass the suggestion request
                        .SuggestUsing(suggestionRequest)
                        .Execute();
                    #endregion
                }
                
                using (var session = store.OpenSession())
                {
                    #region suggest_5
                    // Query for suggested terms for single term:
                    // ==========================================
                    
                    Dictionary<string, SuggestionResult> suggestions = session.Advanced
                         // Make a dynamic document-query on collection 'Products'
                        .DocumentQuery<Product>()
                         // Call 'SuggestUsing' 
                        .SuggestUsing(builder => builder
                             // Request to get terms from field 'Name' that are similar to 'chaig' 
                            .ByField(x => x.Name, "chaig"))
                        .Execute();
                    #endregion
                    
                    #region suggest_6
                    // The resulting suggested terms:
                    // ==============================
                    
                    Console.WriteLine("Suggested terms in field 'Name' that are similar to 'chaig':");
                    foreach (string suggestedTerm in suggestions["Name"].Suggestions)
                    {
                        Console.WriteLine("\t{0}", suggestedTerm);
                    }
                    
                    // Suggested terms in field 'Name' that are similar to 'chaig':
                    //     chai
                    //     chang
                    #endregion
                }
                
                using (var session = store.OpenSession())
                {
                    #region suggest_7
                    // Query for suggested terms for multiple terms:
                    // =============================================
                    
                    Dictionary<string, SuggestionResult> suggestions = session
                         // Make a dynamic query on collection 'Products'
                        .Query<Product>()
                         // Call 'SuggestUsing'
                        .SuggestUsing(builder => builder
                             // Request to get terms from field 'Name' that are similar to 'chaig' OR 'tof' 
                            .ByField(x => x.Name, new[] { "chaig", "tof" }))
                        .Execute();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region suggest_8
                    // Query for suggested terms for multiple terms:
                    // =============================================
                    
                    Dictionary<string, SuggestionResult> suggestions = await asyncSession
                         // Make a dynamic query on collection 'Products'
                        .Query<Product>()
                         // Call 'SuggestUsing' 
                        .SuggestUsing(builder => builder
                             // Request to get terms from field 'Name' that are similar to 'chaig' OR 'tof'
                            .ByField(x => x.Name, new[] { "chaig", "tof" }))
                        .ExecuteAsync();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region suggest_9
                    // Define the suggestion request for multiple terms
                    var suggestionRequest = new SuggestionWithTerms("Name")
                    {
                        // Looking for terms from field 'Name' that are similar to terms 'chaig' OR 'tof'  
                        Terms = new[] { "chaig", "tof"}
                    };
                    
                    // Query for suggestions 
                    Dictionary<string, SuggestionResult> suggestions = session
                        .Query<Product>()
                         // Call 'SuggestUsing' - pass the suggestion request
                        .SuggestUsing(suggestionRequest)
                        .Execute();
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region suggest_10
                    // Query for suggested terms for multiple terms:
                    // =============================================
                    
                    Dictionary<string, SuggestionResult> suggestions = session.Advanced
                         // Make a dynamic document-query on collection 'Products'
                        .DocumentQuery<Product>()
                         // Call 'SuggestUsing'
                        .SuggestUsing(builder => builder
                             // Request to get terms from field 'Name' that are similar to 'chaig' OR 'tof' 
                            .ByField(x => x.Name, new[] { "chaig", "tof" }))
                        .Execute();
                    #endregion
                    
                    #region suggest_11
                    // The resulting suggested terms:
                    // ==============================
                    
                    // Suggested terms in field 'Name' that are similar to 'chaig' or to 'tof':
                    //     chai
                    //     chang
                    //     tofu
                    #endregion
                }

                using (var session = store.OpenSession())
                {
                    #region suggest_12
                    // Query for suggested terms in multiple fields:
                    // =============================================
                    
                    Dictionary<string, SuggestionResult> suggestions = session
                         // Make a dynamic query on collection 'Companies'
                        .Query<Company>()
                         // Call 'SuggestUsing' to get suggestions for terms that are 
                         // similar to 'chaig' in first document field (e.g. 'Name') 
                        .SuggestUsing(builder => builder
                            .ByField(x => x.Name, "chop-soy china"))
                         // Call 'AndSuggestUsing' to get suggestions for terms that are 
                         // similar to 'maria larson' in an additional field (e.g. 'Contact.Name')
                        .AndSuggestUsing(builder => builder
                            .ByField(x => x.Contact.Name, "maria larson"))
                        .Execute();
                    #endregion
                }

                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region suggest_13
                    // Query for suggested terms in multiple fields:
                    // =============================================
                    
                    Dictionary<string, SuggestionResult> suggestions = await asyncSession
                         // Make a dynamic query on collection 'Companies'   
                        .Query<Company>()
                         // Call 'SuggestUsing' to get suggestions for terms that are 
                         // similar to 'chaig' in first document field (e.g. 'Name') 
                        .SuggestUsing(builder => builder
                            .ByField(x => x.Name, "chop-soy china"))
                         // Call 'AndSuggestUsing' to get suggestions for terms that are 
                         // similar to 'maria larson' in an additional field (e.g. 'Contact.Name')
                        .AndSuggestUsing(builder => builder
                            .ByField(x => x.Contact.Name, "maria larson"))
                        .ExecuteAsync();
                    #endregion
                }
                
                using (var session = store.OpenSession())
                {
                    #region suggest_14
                    // Define suggestion requests for multiple fields:
                    
                    var request1 = new SuggestionWithTerm("Name")
                    {
                        // Looking for terms from field 'Name' that are similar to 'chop-soy china'  
                        Term = "chop-soy china"
                    };
                    
                    var request2 = new SuggestionWithTerm("Contact.Name")
                    {
                        // Looking for terms from nested field 'Contact.Name' that are similar to 'maria larson'  
                        Term = "maria larson"
                    };
                    
                    Dictionary<string, SuggestionResult> suggestions = session
                        .Query<Company>()
                         // Call 'SuggestUsing' - pass the suggestion request for the first field
                        .SuggestUsing(request1)
                         // Call 'AndSuggestUsing' - pass the suggestion request for the second field
                        .AndSuggestUsing(request2)
                        .Execute();
                    #endregion
                }
                
                using (var session = store.OpenSession())
                {
                    #region suggest_15
                    // Query for suggested terms in multiple fields:
                    // =============================================
                    
                    Dictionary<string, SuggestionResult> suggestions = session.Advanced
                         // Make a dynamic document-query on collection 'Companies'
                        .DocumentQuery<Company>()
                         // Call 'SuggestUsing' to get suggestions for terms that are 
                         // similar to 'chaig' in first document field (e.g. 'Name') 
                        .SuggestUsing(builder => builder
                            .ByField(x => x.Name, "chop-soy china"))
                         // Call 'AndSuggestUsing' to get suggestions for terms that are 
                         // similar to 'maria larson' in an additional field (e.g. 'Contact.Name')
                        .AndSuggestUsing(builder => builder
                            .ByField(x => x.Contact.Name, "maria larson"))
                        .Execute();
                    #endregion
                    
                    #region suggest_16
                    // The resulting suggested terms:
                    // ==============================
                    
                    // Suggested terms in field 'Name' that is similar to 'chop-soy china':
                    //     chop-suey chinese
                    
                    // Suggested terms in field 'Contact.Name' that are similar to 'maria larson':
                    //     maria larsson
                    //     marie bertrand
                    //     aria cruz
                    //     paula wilson
                    //     maria anders
                    #endregion
                }
                
                using (var session = store.OpenSession())
                {
                    #region suggest_17
                    // Query for suggested terms - customize options and display name:
                    // ===============================================================
                    
                    Dictionary<string, SuggestionResult> suggestions = session
                         // Make a dynamic query on collection 'Products'
                        .Query<Product>()
                         // Call 'SuggestUsing'
                        .SuggestUsing(builder => builder
                            .ByField(x => x.Name, "chaig")
                             // Customize suggestions options
                            .WithOptions(new SuggestionOptions
                            {
                                Accuracy = 0.4f,
                                PageSize = 5,
                                Distance = StringDistanceTypes.JaroWinkler,
                                SortMode = SuggestionSortMode.Popularity
                            })
                             // Customize display name for results
                            .WithDisplayName("SomeCustomName"))
                        .Execute();
                    #endregion
                }
                
                using (var asyncSession = store.OpenAsyncSession())
                {
                    #region suggest_18
                    // Query for suggested terms - customize options and display name:
                    // ===============================================================

                    Dictionary<string, SuggestionResult> suggestions = await asyncSession
                         // Make a dynamic query on collection 'Products'
                        .Query<Product>()
                         // Call 'SuggestUsing'
                        .SuggestUsing(builder => builder
                            .ByField(x => x.Name, "chaig")
                             // Customize suggestions options
                            .WithOptions(new SuggestionOptions
                            {
                                Accuracy = 0.4f,
                                PageSize = 5,
                                Distance = StringDistanceTypes.JaroWinkler,
                                SortMode = SuggestionSortMode.Popularity
                            })
                             // Customize display name for results
                            .WithDisplayName("SomeCustomName"))
                        .ExecuteAsync();
                    #endregion
                }
                
                using (var session = store.OpenSession())
                {
                    #region suggest_19
                    // Define the suggestion request
                    var suggestionRequest = new SuggestionWithTerm("Name")
                    {
                        // Looking for terms from field 'Name' that are similar to term 'chaig'  
                        Term = "chaig",
                        // Customize options
                        Options = new SuggestionOptions 
                        {
                            Accuracy = 0.4f,
                            PageSize = 5,
                            Distance = StringDistanceTypes.JaroWinkler,
                            SortMode = SuggestionSortMode.Popularity
                        },
                        // Customize display name
                        DisplayField = "SomeCustomName"
                    };
                    
                    // Query for suggestions 
                    Dictionary<string, SuggestionResult> suggestions = session
                        .Query<Product>()
                        // Call 'SuggestUsing' - pass the suggestion request
                        .SuggestUsing(suggestionRequest)
                        .Execute();
                    #endregion
                }
                
                using (var session = store.OpenSession())
                {
                    #region suggest_20
                    // Query for suggested terms - customize options and display name:
                    // ===============================================================
                    
                    Dictionary<string, SuggestionResult> suggestions = session.Advanced
                         // Make a dynamic query on collection 'Products'
                        .DocumentQuery<Product>()
                         // Call 'SuggestUsing'
                        .SuggestUsing(builder => builder
                            .ByField(x => x.Name, "chaig")
                             // Customize suggestions options
                            .WithOptions(new SuggestionOptions
                            {
                                Accuracy = 0.4f,
                                PageSize = 5,
                                Distance = StringDistanceTypes.JaroWinkler,
                                SortMode = SuggestionSortMode.Popularity
                            })
                             // Customize display name for results
                            .WithDisplayName("SomeCustomName"))
                        .Execute();
                    #endregion
                    
                    #region suggest_21
                    // The resulting suggested terms:
                    // ==============================
                    
                    Console.WriteLine("Suggested terms:");
                    // Results are available under the custom name entry
                    foreach (string suggestedTerm in suggestions["SomeCustomName"].Suggestions)
                    {
                        Console.WriteLine("\t{0}", suggestedTerm);
                    }
                    
                    // Suggested terms:
                    //     chai
                    //     chang
                    //     chartreuse verte
                    #endregion
                }
            }
        }

        private interface IFoo<T>
        {
            #region syntax_1
            // Overloads for requesting suggestions for term(s) in a field: 
            ISuggestionQuery<T> SuggestUsing<T>(SuggestionBase suggestion);
            ISuggestionQuery<T> SuggestUsing<T>(Action<ISuggestionBuilder<T>> builder);

            // Overloads requesting suggestions for term(s) in another field in the same query:
            ISuggestionQuery<T> AndSuggestUsing(SuggestionBase suggestion);
            ISuggestionQuery<T> AndSuggestUsing(Action<ISuggestionBuilder<T>> builder);
            #endregion

            #region syntax_2
            ISuggestionOperations<T> ByField(string fieldName, string term);
            ISuggestionOperations<T> ByField(string fieldName, string[] terms);
            ISuggestionOperations<T> ByField(Expression<Func<T, object>> path, string term);
            ISuggestionOperations<T> ByField(Expression<Func<T, object>> path, string[] terms);

            ISuggestionOperations<T> WithOptions(SuggestionOptions options);
            ISuggestionOperations<T> WithDisplayName(string displayName);
            #endregion
        }
        
        private class Foo
        {
            #region syntax_3
            public int PageSize { get; set; }
            public StringDistanceTypes? Distance { get; set; }
            public float? Accuracy { get; set; }
            public SuggestionSortMode SortMode { get; set; }
            #endregion
        }
    }
}

File: Documentation/5.4/Samples/python/ClientApi/Session/Querying/HowToWorkWithSuggestions.py
from typing import Union, Callable, TypeVar, List

from ravendb.documents.queries.suggestions import (
    SuggestionWithTerm,
    SuggestionWithTerms,
    SuggestionOptions,
    StringDistanceTypes,
    SuggestionSortMode,
    SuggestionDocumentQuery,
    SuggestionBase,
    SuggestionBuilder,
    SuggestionOperations,
)
from ravendb.infrastructure.operations import CreateSampleDataOperation
from ravendb.infrastructure.orders import Product, Company
from ravendb.serverwide.operations.common import DeleteDatabaseOperation

from examples_base import ExampleBase

_T = TypeVar("_T")


class HowToWorkWithSuggestions(ExampleBase):
    def setUp(self):
        super().setUp()
        with self.embedded_server.get_document_store("Suggestions") as store:
            with store.open_session() as session:
                store.maintenance.send(CreateSampleDataOperation())

    def tearDown(self):
        super().tearDown()
        with self.embedded_server.get_document_store("Suggestions") as store:
            store.maintenance.send(DeleteDatabaseOperation("Suggestions", hard_delete=True))

    def test_how_to_work_with_suggestions(self):
        with self.embedded_server.get_document_store("Suggestions") as store:
            with store.open_session() as session:
                # region suggest_1
                # This dynamic query on the 'Products' collection has NO documents
                products = list(session.query(object_type=Product).where_equals("name", "chaig"))
                # endregion

            with store.open_session() as session:
                # region suggest_2
                # Query for suggested terms for single term:
                # ==========================================
                suggestions = (
                    session.query(object_type=Product)
                    .suggest_using(lambda builder: builder.by_field("name", "chaig"))
                    .execute()
                )
                # endregion

            with store.open_session() as session:
                # region suggest_4
                # Define the suggestion request for single term
                suggestion_request = SuggestionWithTerm("name")
                suggestion_request.term = "chaig"

                # Query for suggestions
                suggestions = (
                    session.query(object_type=Product)
                    # Call 'suggest_using' - pass the suggestion request
                    .suggest_using(suggestion_request).execute()
                )

                # endregion

            with store.open_session() as session:
                # region suggest_5
                # Query for suggested terms for single term:
                # ==========================================
                suggestions = (
                    session.advanced
                    # Make a dynamic document-query on collection 'Products'
                    .document_query(object_type=Product)
                    # Call 'SuggestUsing'
                    .suggest_using(
                        lambda builder: builder
                        # Request to get terms from field 'name' that are similar to 'chaig'
                        .by_field("name", "chaig")
                    ).execute()
                )
                # endregion

                # region suggest_6
                # The resulting suggested terms:
                # ==============================

                print("Suggested terms in field 'name' that are similar to 'chaig':")
                for suggested_term in suggestions["name"].suggestions:
                    print(f"\t{suggested_term}")

                #  Suggested terms in field 'Name' that are similar to 'chaig':
                #  chai
                #  chang
                # endregion

            with store.open_session() as session:
                # region suggest_7
                # Query for suggested terms for multiple terms:
                # =============================================

                suggestions = (
                    session
                    # Make a dynamic query on collection 'Products'
                    .query(object_type=Product)
                    # Call 'suggest_using'
                    .suggest_using(
                        lambda builder: builder
                        # Request to get terms from field 'name' that are similar to 'chaig' OR 'tof'
                        .by_field("name", ["chaig", "tof"])
                    ).execute()
                )
                # endregion
            with store.open_session() as session:
                # region suggest_9
                # Define the suggestion request for multiple terms
                suggestion_request = SuggestionWithTerms("name")
                # Looking for terms from field 'name' that are similar to 'chaig' OR 'tof'
                suggestion_request.terms = ["chaig", "tof"]

                # Query for suggestions
                suggestions = (
                    session.query(object_type=Product)
                    # Call 'suggest_using' - pass the suggestion request
                    .suggest_using(suggestion_request).execute()
                )

                # endregion
            with store.open_session() as session:
                # region suggest_10
                #  Query for suggested terms for multiple terms:
                #  =============================================

                suggestions = (
                    session.advanced
                    # Make a dynamic document-query on collection 'Products'
                    .document_query(object_type=Product)
                    # Call 'suggest_using'
                    .suggest_using(
                        lambda builder: builder
                        # Request to get terms from field 'name' that are similar to 'chaig' OR 'tof'
                        .by_field("name", ["chaig", "tof"])
                    )
                ).execute()
                # endregion
                # region suggest_11
                # The resulting suggested terms:
                #  ==============================
                #
                # Suggested terms in field 'Name' that are similar to 'chaig' OR to 'tof':
                #      chai
                #      chang
                #      tofu
                # endregion
            with store.open_session() as session:
                # region suggest_12
                # Query for suggested terms in multiple fields:
                # =============================================

                suggestions = (
                    session
                    # Make a dynamic query on collection 'Companies'
                    .query(object_type=Company)
                    # Call 'suggest_using' to get suggestions for terms that are
                    # similar to 'chop-soy china' in first document field (e.g. 'name')
                    .suggest_using(lambda builder: builder.by_field("name", "chop-soy china"))
                    # Call 'and_suggest_using' to get suggestions for terms that are
                    # similar to 'maria larson' in an additional field (e.g. 'Contact.Name')
                    .and_suggest_using(lambda builder: builder.by_field("contact.name", "maria larson")).execute()
                )
                # endregion
            with store.open_session() as session:
                # region suggest_14
                # Define suggestion requests for multiple fields:

                request1 = SuggestionWithTerm("name")
                # Looking for terms from field 'Name' that are similar to 'chop-soy china'
                request1.term = "chop-soy china"

                request2 = SuggestionWithTerm("contact.name")
                # Looking for terms from nested field 'Contact.Name' that are similar to 'maria larson'
                request2.term = ["maria larson"]

                suggestions = (
                    session.query(object_type=Company)
                    # Call 'suggest_using' - pass the suggestion request for the first field
                    .suggest_using(request1)
                    # Call 'and_suggest_using' - pass the suggestion request for the second field
                    .and_suggest_using(request2).execute()
                )
                # endregion
            with store.open_session() as session:
                # region suggest_15
                # Query for suggested terms in multiple fields:
                # =============================================
                suggestions = (
                    session.advanced
                    # Make a dynamic document-query on collection 'Companies'
                    .document_query(object_type=Company)
                    # Call 'suggest_using' to get suggestions for terms that are
                    # similar to 'chop-soy china' in first document field (e.g. 'name')
                    .suggest_using(lambda builder: builder.by_field("name", "chop-soy china"))
                    # Call 'and_suggest_using' to get suggestions for terms that are
                    # similar to 'maria larson' in an additional field (e.g. 'Contact.Name')
                    .and_suggest_using(lambda builder: builder.by_field("contact.name", "maria larson")).execute()
                )
                # endregion
                # region suggest_16
                # The resulting suggested terms:
                # ==============================
                #
                # Suggested terms in field 'name' that is similar to 'chop-soy china':
                #     chop-suey chinese
                #
                # Suggested terms in field 'contact.name' that are similar to 'maria larson':
                #     maria larsson
                #     marie bertrand
                #     aria cruz
                #     paula wilson
                #     maria anders
                # endregion
            with store.open_session() as session:
                # region suggest_17
                #  Query for suggested terms - customize options and display name:
                #  ===============================================================
                suggestions = (
                    session
                    # Make a dynamic query on collection 'Products'
                    .query(object_type=Product)
                    # Call 'suggest_using'
                    .suggest_using(
                        lambda builder: builder.by_field("name", "chaig")
                        # Customize suggestion options
                        .with_options(
                            SuggestionOptions(
                                accuracy=0.4,
                                page_size=5,
                                distance=StringDistanceTypes.JARO_WINKLER,
                                sort_mode=SuggestionSortMode.POPULARITY,
                            )
                        )
                        # Customize display name for results
                        .with_display_name("SomeCustomName")
                    ).execute()
                )
                # endregion
            with store.open_session() as session:
                # region suggest_19
                # Define the suggestion request
                suggestion_request = SuggestionWithTerm("name")
                # Looking for terms from field 'Name' that are similar to term 'chaig'
                suggestion_request.term = "chaig"
                # Customize options
                suggestion_request.options = SuggestionOptions(
                    accuracy=5,
                    page_size=5,
                    distance=StringDistanceTypes.JARO_WINKLER,
                    sort_mode=SuggestionSortMode.POPULARITY,
                )
                # Customize display name
                suggestion_request.display_field = "SomeCustomName"

                # Query for suggestions
                suggestions = (
                    session.query(object_type=Product)
                    # Call 'suggest_using' - pass the suggestion request
                    .suggest_using(suggestion_request).execute()
                )
                # endregion
            with store.open_session() as session:
                # region suggest_20
                # Query for suggested terms - customize options and display name:
                # ===============================================================
                suggestions = (
                    session.advanced
                    # Make a dynamic query on collection 'Products'
                    .document_query(object_type=Product)
                    # Call 'suggest_using'
                    .suggest_using(
                        lambda builder: builder.by_field("name", "chaig")
                        # Customize suggestion options
                        .with_options(
                            SuggestionOptions(
                                accuracy=0.4,
                                page_size=5,
                                distance=StringDistanceTypes.JARO_WINKLER,
                                sort_mode=SuggestionSortMode.POPULARITY,
                            )
                        )
                        # Customize display name for results
                        .with_display_name("SomeCustomName")
                    ).execute()
                )
                # endregion
                # region suggest_21
                # The resulting suggested terms:
                # ==============================

                print("Suggested terms:")
                # Results are available under the custom name entry
                for suggested_term in suggestions["SomeCustomName"].suggestions:
                    print(f"\t{suggested_term}")

                # Suggested terms:
                #     chai
                #     chang
                #     chartreuse verte
                # endregion


class Foo:
    # region syntax_1
    # Method for requesting suggestions for term(s) in a field:
    def suggest_using(
        self, suggestion_or_builder: Union[SuggestionBase, Callable[[SuggestionBuilder[_T]], None]]
    ) -> SuggestionDocumentQuery[_T]: ...

    # Method for requesting suggestions for term(s) in another field in the same query:
    def and_suggest_using(
        self, suggestion_or_builder: Union[SuggestionBase, Callable[[SuggestionBuilder[_T]], None]]
    ) -> SuggestionDocumentQuery[_T]: ...

    # endregion

    # region syntax_2
    def by_field(self, field_name: str, term_or_terms: Union[str, List[str]]) -> SuggestionOperations[_T]: ...

    def with_display_name(self, display_name: str) -> SuggestionOperations[_T]: ...
    def with_options(self, options: SuggestionOptions) -> SuggestionOperations[_T]: ...

    # endregion


class Foo2:
    # region syntax_3
    DEFAULT_ACCURACY = 0.5
    DEFAULT_PAGE_SIZE = 15
    DEFAULT_DISTANCE = StringDistanceTypes.LEVENSHTEIN
    DEFAULT_SORT_MODE = SuggestionSortMode.POPULARITY

    def __init__(
        self,
        page_size: int = DEFAULT_PAGE_SIZE,
        distance: StringDistanceTypes = DEFAULT_DISTANCE,
        accuracy: float = DEFAULT_ACCURACY,
        sort_mode: SuggestionSortMode = DEFAULT_SORT_MODE,
    ):
        self.page_size = page_size
        self.distance = distance
        self.accuracy = accuracy
        self.sort_mode = sort_mode

    # endregion
